### DISCUSSION TITLE
SAML2 based authrization workaround

### DISCUSSION THREAD

**Discussion by yamsania** [uwrit/leaf#459] [Q&A]
Hey @ndobb ,

We recently implemented Leaf on GCP and are trying to setup authorization and are failing to do so, as we use third party IDP and the IDP does not have group information as its not synced from active directory. What would be our options if we cannot get group information from SAML2? 
@glabrie10 Please help us answer any questions @ndobb may have for us.

**Comment by mh2727** [comment]
Hi @yamsania.

My name is Mehadi Hassan and I am one of the developers on the Leaf team.

Re: `What would be our options if we cannot get group information from SAML2?`
What do you mean by you're unable to get group information? Does the IDP provide info on roles or is there a way to categorize users?

Leaf is looking for some identifier that can differentiate users based on their assigned role, group, or whatever your team decides to use to categorize users. As long as that identifier is sent, Leaf will be able to use it to grant access.

You should be able to manage users in your IDP and group them with the necessary permissions. Whatever you use to group the users is your group information.

--
Mehadi Hassan

**Reply by yamsania** [reply]
Hey @mh2727,

The thing is that we use thirt party IDP for saml and unfortunetly the group information is not being synced to the third party idp. which leaves us getting only the user information from metadata but not the group information. so wondering if there is a way to manage groups locally on the application server?

**Reply by mh2727** [reply]
Because Leaf reads the SAML headers and assesses the permissions based on that, there's not a way to locally manage groups unless it's sent over as SAML headers. Additionally, you can manage access to Leaf with Apache/Nginx/etc., but that won't carry over to the application server and satisfy Leaf's groups needs.

I'd recommend reaching out to the IDP and figuring out what the issue is. There should be a way to configure groups/roles that can be sent over in the metadata. Or looking into finding another identifier that can be used to manage users.

--
Mehadi Hassan

**Reply by glabrie10** [reply]
Hey @mh2727

Gabe here from the application team with Health Data Compass.   

Unfortunately, our IPD cannot do groups.  We are able to create other identifiers for entitlement attributes.  What would we have to change in the appsetting.json configuration to use our custom entitlement attributes instead of a group?


Gabriel LaBrie 
Cloud Engineer
Health Data Compass

**Reply by mh2727** [reply]
Hi @glabrie10 

Under `Authorization -> SAML2 -> HeadersMapping -> Entitlements`, you can specify the `Name` of the entitlement you want to use and the `Delimiter`. This will allow Leaf to read that header.

Then, you can specify the different roles (`Authorization -> SAML2 -> RolesMapping`) based on your entitlement.

Here is an example to help you get started.

```
The headers sent over SAML:

    example_entitlement: groupA; groupB; groupC

    The `Name` is "example_entitlement".
    The `Delimiter` is ";".

The different roles would be specified like so:

    "RolesMapping": {
        "User": "groupA",
        "Super": "groupB",
        "Identified": "groupC",
        "Admin": "groupC",
        "Federated": "groupA"
      }
```

You can read more about this on our Leaf docs page: https://leafdocs.rit.uw.edu/installation/installation_steps/6_appsettings/#authorization

Hope this helps and answers your questions.

--
Mehadi Hassan

**Reply by glabrie10** [reply]
@mh2727 

Thank you for the input.  We were able to use custom identifiers but we still can't get authorization to work.  Would it be possible to schedule a zoom and go over what we are trying?

**Reply by mh2727** [reply]
Hi @glabrie10 

I just sent an email requesting some times that work for you next week.

--
Mehadi Hassan

**Reply by glabrie10** [reply]
Thanks, Mehadi!

**Comment by artgoldberg** [comment]
Folks. See #436 which may solve your problem and has been implemented.

**Comment by ndobb** [comment]
@artgoldberg, thanks for pointing that out.

On a related note @yamsania @glabrie10, I've just released v3.10, which [add support for user authorization via the Leaf app DB](https://github.com/uwrit/leaf/releases/tag/v3.10.0), as well as the feature for allowing all users, which @artgoldberg mentioned. 

You should be able to use your IdP to authenticate the user (and get their username), then the app DB tables to manage their roles if you'd like.

Let me know if you have any questions.

**Reply by artgoldberg** [reply]
@ndobb I like the idea of combining SAML authentication with LeafDB authorization.
